zkvm/src/contract.rs

```diff
@@ -0,0 +1,164 @@
+use crate::encoding;
+use crate::encoding::Subslice;
+use crate::errors::VMError;
+use crate::predicate::Predicate;
+use crate::txlog::{TxID, UTXO};
+use crate::types::{Commitment, Data, Value};
+
+/// Prefix for the data type in the Output Structure
+pub const DATA_TYPE: u8 = 0x00;
+
+/// Prefix for the value type in the Output Structure
+pub const VALUE_TYPE: u8 = 0x01;
+
+/// A ZkVM contract that holds a _payload_ (a list of portable items) protected by a _predicate_.
+#[derive(Debug)]
+pub struct Contract {
+    pub(crate) payload: Vec<PortableItem>,
+    pub(crate) predicate: Predicate,
+}
+
+/// Representation of items that can be stored within outputs and contracts.
+#[derive(Debug)]
+pub enum PortableItem {
+    Data(Data),
+    Value(Value),
+}
+
+#[derive(Clone, Debug)]
+pub struct Input {
+    pub contract: FrozenContract,
+    pub utxo: UTXO,
```

Comment:
lets make all of these at least `pub(crate)` because we don't want to allow constructing an Input manually and making inconsistent utxo ID w.r.t. txid and contract.